Can't figure it out.

MartinCernohorsky opened the pull request:
Can someone take a look at ProductService.svc.cs and its TODO? How can I detect product was changed if I change its property which happens to be referential type. Due to debugger it has been changed just EntityState stays as is.

Commit:
Can't figure it out.

mozdren commented on ServiceBus/ServiceBus/IProductService.cs:
```diff
@@ -27,32 +27,39 @@ public interface IProductService
         /// <returns>Result object</returns>
         [OperationContract (Name = "AddProduct")]
-        SharedLibs.DataContracts.Result AddProduct(string name, double price, Guid guid);
+        SharedLibs.DataContracts.Result AddProduct(string name, double price, Guid guid, int pType, bool headliner);
```
1) Please, use whole names for parameters pType -> productType.
2) Why is product type an integer? I would rather see enumeration.

mozdren commented on ServiceBus/SharedLibs/DataContracts/Product.cs:
```diff
@@ -6,6 +6,7 @@ namespace SharedLibs.DataContracts
     /// <summary>
     /// Datacontract for product
     /// </summary>
+    //TODO Is obsolote datacontract or not? 
```
The data contract is not obsolete, but has to be extended so we provide the client with all relevant information (that is also true for all other data contracts)

MartinCernohorsky replied:
Could you please, as a business, let me know what do you want to get inside this DTO ? Especially those collections concerns me.  (I do not want to return too much or too little and I do not know the optimal ratio :( )

mozdren replied:
We should provide the customer with all possible information. Only internal (our) data should not be shared. Right now we have no internal information stored in products, therefore we can add everything.

mozdren commented on ServiceBus/ServiceBusTests/ProductServiceTests.cs:
```diff
@@ -1,15 +1,18 @@
     [TestClass]
     public class ProductServiceTests
     {
+        ServiceBusTests.ProductService.ProductServiceClient client = new ProductService.ProductServiceClient();
```
I know that this looks better, but in case of tests (which serve like documentation and help for developers) the client should be created in each test method separately.

mozdren commented on ServiceBus/ServiceBus/ProductService.svc.cs:
```diff
@@ -136,63 +141,74 @@ public Result AddProduct(string name, double price, Guid guid)
         {
             try
             {
-                var originalProduct = GetProduct(guid);
```
this method should be much simpler. It should:
1) get the product using the DB context
2) change the values in retrieved product (instance from the db context)
3) return new instance of data contract "ServiceBus.DataContracts.Product" representing the changed item

MartinCernohorsky replied:
Well this is a previous version. Actual works with db context but it does not work when I try to detect whether changes have been made or not.

mozdren reviewed (changes requested):
If you want me to change the code, then I can do that on Saturday or Sunday (not earlier)

MartinCernohorsky commented:
I'll try to fix most of it according to your advices today.

mozdren commented on ServiceBus/ServiceBus/ProductService.svc.cs:
```diff
@@ -136,63 +141,74 @@ public Result AddProduct(string name, double price, Guid guid)
             {
-                var originalProduct = GetProduct(guid);
-
-                if (originalProduct.Result.ResultType == ResultType.Success)
```
originalProduct.Result.IsSuccess doesn't work?

MartinCernohorsky replied:
I have changed this part because method get guid then it search for DTO product then new object (context) for product is created then it is populated with data from DTO product object and then some modification could begin.

Now it should be get guid, find guid in db and if found create object filled with database data.

Goal was to simplify the method.

mozdren commented on ServiceBus/ServiceBus/ProductService.svc.cs:
```diff
@@ -136,63 +141,74 @@ public Result AddProduct(string name, double price, Guid guid)
-
                     using (var context = new ServiceBusDatabaseEntities())
                     {
-                        context.Products.Attach(editableProduct);
```
I am not sure if this is a correct way how to handle this... I would simply take data from parameters of the method, compare it with object retrieved from context and then update the product. I find this code over complicated.

MartinCernohorsky replied:
It works when value types of the objects are changed. Referential are ignored. 

I chosen this way because I wanted to distinguish a scenario when someone want to change something but input parameters are exactly the same as those which are already present in db. I can try something like 

bool flag = false;

if(att is valid){ ... ; flag = true;}
if(att2 is valid) {... ; flag = true;}
...
if (flag){
saveChanges(); 
}
else{ Result("nothing has been changed") }

mozdren commented on ServiceBus/ServiceBus/Mock/ProductService.svc.cs:
```diff
@@ -24,8 +26,10 @@ public Result AddProduct(SharedLibs.DataContracts.Product product)
+        /// <param name="headliner">Is this product supposed to be main product?</param>
         /// <returns>success if added successfully</returns>
-        public Result AddProduct(string name, double price, Guid guid)
+        public Result AddProduct(string name, double price, Guid guid, int pType, bool headliner= false)
```
formating of the "="

Commit:
New DTO objects

New DTOs created solely to mirror context model.

There is also enum ProductTypes created due to comment in previous
commit. I do not know how to benifit from it now but I would like to
leave it there if you don't mind.

Commit:
Revert "New DTO objects"

This reverts commit a85027516c811cf087c20a3ae55bc2a62e1a1b59.

Commit:
Revert "Revert "New DTO objects""

This reverts commit 7bd351034c240b28cd76f00c1f112b5ca1a26af1.

Commit:
Finish till end of the week if possible.
